### DISCUSSION TITLE
<NumberInput> with react-hook-form

### DISCUSSION THREAD

**Discussion by ricardocanelas** [chakra-ui/chakra-ui#2496] [General] [answered]
Does anybody have an example of how to use NumberInput with `react-hook-form@ˆ6.0`?

**Comment by ricardocanelas** [comment]
My solution was...

```js
<NumberInput
  step={1}
  defaultValue={18}
  ref={register({
    name: 'age',
    required: true
  })}
  onChange={(valueString) => setValue('age', valueString)}
>
  <NumberInputField name='age' />
</NumberInput>
```

**Reply by RealDyllon** [reply]
Hey @ricardocanelas, is there a solution for this that works with Typescript?

**Reply by nakaakist** [reply]
https://github.com/chakra-ui/chakra-ui/issues/2832#issuecomment-748390514

I tried this solution, and it perfectly worked with TypeScript.

**Comment by xuanvan229** [comment, marked as answer]
Anyone want a solution for Typescript. Here is it 

```js
<Controller
  name={"price"}
  control={control}
  rules={{
    required: {
      value: true,
      message: "Price is required",
    },
  }}
  render={({ field: { ref, ...restField } }) => (
    <NumberInput {...restField} min={0}>
      <NumberInputField ref={ref} name={restField.name} />
    </NumberInput>
  )}
```

**Reply by JayV30** [reply]
This is the best answer and should be the accepted answer! I first tried the accepted answer as it was faster to implement, but the `useForm` reset function wasn't working properly and some level of control from the `useForm` hook was just not being passed to the input. Well, clearly that's because the ref wasn't being passed properly. Your solution properly passes the ref and everything works.

I also had to modify the ref on the `NumberInputField` because I also needed it to be the first focus ref in a popup.  

For anyone reading in the future, here's what my implementation looks like:

````
<Controller
  name="hours"
  control={control}
  render={({ field: { ref, ...restField } }) => (
    <NumberInput
      {...restField}
      precision={2}
      step={0.25}
      min={0.25}
      max={24}
    >
      <NumberInputField
        ref={(node) => {
          ref.current = node;
          firstFieldRef.current = node;
        }}
        name={restField.name}
      />
      <NumberInputStepper>
        <NumberIncrementStepper />
        <NumberDecrementStepper />
      </NumberInputStepper>
    </NumberInput>
  )}
/>
````

**Comment by gmoqa** [comment]
That's how I did it

```html
<NumberInput min={0} max={100}>
    <NumberInputField {...register('percentage', {
        required: 'This field is required'
    })}/>
</NumberInput>
```


Saludos

**Comment by andrehadianto** [comment]
I can't seem to make it work when using `useNumberInput()` hook. Please help 🙏

**Reply by andrehadianto** [reply]
oh, turns out you only need to replace the `onChange` param from the `useNumberInput` hook implementation
```
const { getIncrementButtonProps, getDecrementButtonProps, getInputProps } =
  useNumberInput({
    onChange: (_, valueAsNumber) => props.onChange(valueAsNumber as any),
  });
  ```
whereas `props`  passes `field['onChange']` from the parent component

**Comment by wnharper** [comment]
I have been stuck on this all day. For some reason I cannot for the life of me get the form to reset the `NumberInput` correctly. Before I found this thread it wouldn't reset at all. After implementing the solution above, it resets the value to 1. 
If I replace the `NumberInput` component with an Input component using `type="number"` it works fine.

Here is my code:
```typescript


<Controller
      control={control}
      name="price"
      render={({ field: { ref, ...restField } }) => (
        <FormControl
          isInvalid={!!errors.price}
        >
          <NumberInput {...restField} min={0} size="sm">
            <NumberInputField
              ref={ref}
              name={restField.name}
            />
            <NumberInputStepper>
              <NumberIncrementStepper />
              <NumberDecrementStepper />
            </NumberInputStepper>
          </NumberInput>
          <FormErrorMessage>
            {errors.price.message}
          </FormErrorMessage>
        </FormControl>
      )}
    />
```
